**SPAM SMS DETECTION**

by Srijita Ghosh Hajra



Build an AI model that can classify SMS messages as spam or legitimate. Use techniques like TF-IDF or word embeddings with classifiers like Naive Bayes, Logistic Regression, or Support Vector Machines to identify spam messages.

Dataset link: https://www.kaggle.com/datasets/uciml/sms-spam-collection-dataset/download?datasetVersionNumber=1

Importing libraries

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

loading the dataset to a Pandas DataFrame

In [30]:
# loading the data from csv file to a pandas Dataframe
try:
    df = pd.read_csv("spam.csv", encoding='utf-8')
except UnicodeDecodeError:
    df = pd.read_csv("spam.csv", encoding='latin-1')

In [3]:
df

v1                                                 v2 Unnamed: 2  \
0      ham  Go until jurong point, crazy.. Available only ...        NaN   
1      ham                      Ok lar... Joking wif u oni...        NaN   
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3      ham  U dun say so early hor... U c already then say...        NaN   
4      ham  Nah I don't think he goes to usf, he lives aro...        NaN   
...    ...                                                ...        ...   
5567  spam  This is the 2nd time we have tried 2 contact u...        NaN   
5568   ham              Will Ì_ b going to esplanade fr home?        NaN   
5569   ham  Pity, * was in mood for that. So...any other s...        NaN   
5570   ham  The guy did some bitching but I acted like i'd...        NaN   
5571   ham                         Rofl. Its true to its name        NaN   

     Unnamed: 3 Unnamed: 4  
0           NaN        NaN  
1           NaN        NaN  
2           NaN        NaN  
3           NaN        NaN  
4           NaN        NaN  
...         ...        ...  
5567        NaN        NaN  
5568        NaN        NaN  
5569        NaN        NaN  
5570        NaN        NaN  
5571        NaN        NaN  

[5572 rows x 5 columns]

In [4]:
# checking the number of rows and columns in the dataframe
df.shape

(5572, 5)

Label Encoding

Exploratory Data Analysis

In [5]:
df.isnull().values.any()

True

In [6]:
# label spam mail as 0 and ham mail as 1

df.loc[df['v1'] == 'spam', 'v1',] = 0
df.loc[df['v1'] == 'ham', 'v1',] = 1

In [7]:
# separating the data as texts and label

X = df['v2']

Y = df['v1']

In [12]:
print(X.shape,Y.shape)

(5572,) (5572,)


In [14]:
X

0       Go until jurong point, crazy.. Available only ...
1                           Ok lar... Joking wif u oni...
2       Free entry in 2 a wkly comp to win FA Cup fina...
3       U dun say so early hor... U c already then say...
4       Nah I don't think he goes to usf, he lives aro...
                              ...                        
5567    This is the 2nd time we have tried 2 contact u...
5568                Will Ì_ b going to esplanade fr home?
5569    Pity, * was in mood for that. So...any other s...
5570    The guy did some bitching but I acted like i'd...
5571                           Rofl. Its true to its name
Name: v2, Length: 5572, dtype: object

In [13]:
Y

0       1
1       1
2       0
3       1
4       1
       ..
5567    0
5568    1
5569    1
5570    1
5571    1
Name: v1, Length: 5572, dtype: object

In [15]:
X_train

3075    Mum, hope you are having a great day. Hoping t...
1787                           Yes:)sura in sun tv.:)lol.
1614    Me sef dey laugh you. Meanwhile how's my darli...
4304                Yo come over carlos will be here soon
3266                    Ok then i come n pick u at engin?
                              ...                        
789                          Gud mrng dear hav a nice day
968             Are you willing to go for aptitude class.
1667    So now my dad is gonna call after he gets out ...
3321    Ok darlin i supose it was ok i just worry too ...
1688                     Nan sonathaya soladha. Why boss?
Name: v2, Length: 4457, dtype: object

In [16]:
X_test

2632                       I WILL CAL YOU SIR. In meeting
454     Loan for any purpose å£500 - å£75,000. Homeown...
983     LOOK AT THE FUCKIN TIME. WHAT THE FUCK YOU THI...
1282    Ever green quote ever told by Jerry in cartoon...
4610                                  Wat time Ì_ finish?
                              ...                        
4827    Lol no. Just trying to make your day a little ...
5291      Xy trying smth now. U eat already? We havent...
3325    Huh so fast... Dat means u havent finished pai...
3561    Still chance there. If you search hard you wil...
1136    Dont forget you can place as many FREE Request...
Name: v2, Length: 1115, dtype: object

In [17]:
 Y_train

3075    1
1787    1
1614    1
4304    1
3266    1
       ..
789     1
968     1
1667    1
3321    1
1688    1
Name: v1, Length: 4457, dtype: object

In [18]:
Y_test

2632    1
454     0
983     1
1282    1
4610    1
       ..
4827    1
5291    1
3325    1
3561    1
1136    0
Name: v1, Length: 1115, dtype: object

Splitting the data into training data & test data

In [11]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=3)

Feature Extraction

In [19]:
feature_extraction = TfidfVectorizer()
X_train_features = feature_extraction.fit_transform(X_train)
X_test_features = feature_extraction.transform(X_test)

# convert Y_train and Y_test values as integers

Y_train = Y_train.astype('int')
Y_test = Y_test.astype('int')


In [20]:
print(X_train_features)

  (0, 741)	0.28307455118083463
  (0, 3360)	0.1327523238442287
  (0, 4108)	0.21196015023008544
  (0, 4908)	0.12973225055917514
  (0, 3042)	0.26300555749396887
  (0, 946)	0.1151770452043338
  (0, 7464)	0.19261204588580316
  (0, 4431)	0.3421657916670175
  (0, 6805)	0.17846848640014898
  (0, 6873)	0.15216101010779184
  (0, 3497)	0.28307455118083463
  (0, 2178)	0.33952544349598
  (0, 3235)	0.3869898904365042
  (0, 3365)	0.22753426247664568
  (0, 1032)	0.1361275560580212
  (0, 7720)	0.1765620792792692
  (0, 3491)	0.19174855251416806
  (0, 4655)	0.2558426236041184
  (1, 4190)	0.3725861907992424
  (1, 7099)	0.42172200036894236
  (1, 6620)	0.46707907862382136
  (1, 3646)	0.2020333473623602
  (1, 6645)	0.553594666958471
  (1, 7704)	0.3433404875792393
  (2, 954)	0.4257390912308466
  :	:
  (4455, 7402)	0.15130978849620824
  (4455, 6316)	0.16857953235415776
  (4455, 6850)	0.14840315498751144
  (4455, 1592)	0.11611242093594701
  (4455, 6991)	0.1743411711262433
  (4455, 4647)	0.1711510506728716
  (44

Training the Model

Logistic Regression

In [21]:
model = LogisticRegression()

In [22]:
# training the Logistic Regression model with the training data
model.fit(X_train_features, Y_train)

LogisticRegression()

Evaluating the trained model

In [23]:
# prediction on training data

prediction_on_training_data = model.predict(X_train_features)
accuracy_on_training_data = accuracy_score(Y_train, prediction_on_training_data)

In [24]:
print('Accuracy on training data : ', accuracy_on_training_data)

Accuracy on training data :  0.9739735247924612


In [25]:
# prediction on test data

prediction_on_test_data = model.predict(X_test_features)
accuracy_on_test_data = accuracy_score(Y_test, prediction_on_test_data)

In [26]:
print('Accuracy on test data : ', accuracy_on_test_data)

Accuracy on test data :  0.9757847533632287


Building a Predictive System

In [29]:
input_mail = ["WINNER!! As a valued network customer you have been selected to receivea å£900 prize reward! To claim call 09061701461. Claim code KL341. Valid 12 hours only."]

# convert text to feature vectors
input_data_features = feature_extraction.transform(input_mail)

# making prediction

prediction = model.predict(input_data_features)
print(prediction)


if (prediction[0]==1):
  print('Ham mail')

else:
  print('Spam mail')

[0]
Spam mail


Conclusion:


1.   Using Logistic regression, accuracy of training and testing data are 0.9739735247924612 and 0.9757847533632287 respectively.
2. So, we can see that my model performs well using Logistic regression for predict whether mail is spam or not.

